Taking Predef into the modern age.

paulp opened the pull request:
Removed some methods in Predef, deprecated some others, and turned its implicit classes into implicit classes.

Review by @retronym

paulp committed:
Made ArrowAssoc and Ensuring implicit classes.

paulp committed:
Deprecated the read* methods in Predef.

For hopefully obvious reasons. These are little used and
absolutely unnecessary to be in global namespace, especially
methods with weird ad hoc semantics yet tempting names such
as readBoolean().

paulp committed:
Removed more deprecated methods in Predef.

And updated the many tests which finally had to let go of
unqualified exit and error.

paulp commented on src/compiler/scala/tools/nsc/interactive/REPL.scala:
```diff
@@ -189,7 +189,7 @@ object REPL {
         case List("quit") =>
           comp.askShutdown()
-          exit(1) // Don't use sys yet as this has to run on 2.8.2 also.
+          sys.exit(1)
```
In case anyone is wondering, the usage of string interpolation in this same file makes it unlikely this still has to work on 2.8.2.

som-snytt replied:
I won't be able to `exit` the repl anymore with the deprecation message that makes me stop for two or three seconds and think first, _warning_?, and second, what will I do when they take away `exit`.

I quit sbt with `bye`, so probably I will rename on import.

I can't use `:q` because I reflexively `:wq`.  I hope git markup doesn't turn :q into some sort of face.

But will `-Xlint:strict` warn me that I shouldn't use `sys.exit` in case my code winds up in a context that doesn't care to terminate the jvm just now?

paulp replied:
We could bind exit up front in the repl;

```
silently("def exit = $intp.close()")
```

or something.

soc commented:
LGTM!

I really really hope you resurrected my old "remove `exit`" patch instead of recreating it, because I remember it was a lot of pain and I wish nobody has to get through it again. :-)

Btw, this will most likely break SBT for the next few decades, are we prepared to handle that?

paulp commented:
I didn't remember that one existed, but my experience is that due to merge pain a patch like that is no less work than doing it again. I don't see why this would have to break sbt, but if that's the case we'll accommodate it somehow.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1624/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2355/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1639/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2355/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2355/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1639/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

paulp committed:
Address scaladoc test failures.

And restore stub non-implicit versions of former Predef methods
since they need to be deprecated for those cases where the names
were referenced directly.

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1659/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2390/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1659/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2390/ <br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

soc commented:
```
[scalacfork] /localhome/jenkins/d/workspace/pr-scala-testsuite-linux-opt/build/sbt-interface/0.12.0/src/Command.scala:20: error: not found: value error
[scalacfork]            constr(classOf[List[_]], classOf[Settings], classOf[Function1[_, _]], classOf[Boolean]).newInstance(arguments, settings, error _, false.asInstanceOf[AnyRef])
```

paulp commented:
Wonderful, it has only been deprecated since 2.9.0. Not like we could have INLINED that by now.

I wonder if this will cause problems...

```
package object xsbt {
  def error(message: String): Nothing = throw new RuntimeException(message)
}
```

paulp commented:
Superseded by #1999.
